# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/debugger/Source.cpp
# Commit: 554b7637fe60
# Full Hash: 554b7637fe60e8a57e5180b5daaa1441f1bea31b
# Author: Denis Palmeiro <dpalmeiro@mozilla.com>
# Date: 2020-05-13 03:44:05
# Regressor Bug: 1501608
# File Overlap Count: 1
# Description:
#   Bug 1501608 - Remove the ELEMENT_SLOT in the ScriptSourceObject and instead use a callback function to return the script element based on the value of the privateValue in the SSO. r=jonco,smaug
#   
#   We can reduce the size of the SSO by removing the element slot entirely, and instead retrieve the element through a callback function.  The callback will take in the value in the private slot of the SSO, which is either a LoadedScript* (from the browser) or a JSObject* (from the shell).  In addition, this removes the requirement of having a script dom element ready when parsing a JS script which can open up new opportunities for performance.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D70417
# ==============================================================================

diff -r 3470d926c02a -r 554b7637fe60 js/src/debugger/Source.cpp
--- a/js/src/debugger/Source.cpp	Tue May 12 09:40:48 2020 +0000
+++ b/js/src/debugger/Source.cpp	Tue May 12 19:28:15 2020 +0000
@@ -378,15 +378,17 @@
 }
 
 struct DebuggerSourceGetElementMatcher {
+  JSContext* mCx = nullptr;
+  explicit DebuggerSourceGetElementMatcher(JSContext* cx_) : mCx(cx_) {}
   using ReturnType = JSObject*;
   ReturnType match(HandleScriptSourceObject sourceObject) {
-    return sourceObject->unwrappedElement();
+    return sourceObject->unwrappedElement(mCx);
   }
   ReturnType match(Handle<WasmInstanceObject*> wasmInstance) { return nullptr; }
 };
 
 bool DebuggerSource::CallData::getElement() {
-  DebuggerSourceGetElementMatcher matcher;
+  DebuggerSourceGetElementMatcher matcher(cx);
   if (JSObject* element = referent.match(matcher)) {
     args.rval().setObjectOrNull(element);
     if (!Debugger::fromChildJSObject(obj)->wrapDebuggeeValue(cx, args.rval())) {